Generate RegionBackendService in terraform

slevenick opened the pull request:
<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.
-->

<!-- Your regular pull request description goes here -->
Generate RegionBackendService via MM for terraform. There should be no functionality change in the resulting resource. 



<!--
For each repository you expect to modify with this PR, fill in a repo-specific
PR title under the corresponding tag. We use repo-specified PR titles to ensure
that each downstream has a clear, easy to understand history.

If the Magician generates a PR for a repo with no specified title, it will use
the title of this PR. [terraform-beta] will inherit the title of [terraform]
if it has no specified title.
-->

<!-- Optional PR titles that describe your downstream changes -->
-----------------------------------------------------------------
# [all]
## [terraform]
### [terraform-beta]
## [ansible]
## [inspec]

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of 22177c385fd13fb566112bfd7a79150c6d296d02. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/645 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/65 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/3521

slevenick commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,266 @@ objects:
+        properties:
+          - !ruby/object:Api::Type::Integer
+            name: 'drainingTimeoutSec'
+            default_value: 0
```
This defaults to 0 to keep with current functionality

rileykarson replied:
We can do this as an override in TF so Ansible / Inspec infer correct values later on if you'd like as well.

slevenick commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,266 @@ objects:
+        description: |
+          The protocol this BackendService uses to communicate with backends.
+
+          The possible values are TCP and UDP, and the default is TCP.
```
Do we prefer to allow the API to default this or default via `default_value`?

rileykarson replied:
We should set a default_value and then not use it in Terraform I think- adding one probably counts as a breaking change.

slevenick replied:
Is it possible to override a default_value to remove it?

rileykarson replied:
No idea! If it's impossible, I'd comment out the `default_value` in the api.yaml and note that downstreams need to implement it as an override

slevenick commented on products/compute/terraform.yaml:
```diff
@@ -233,6 +233,68 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         diff_suppress_func: 'compareSelfLinkOrResourceName'
       timeoutSec: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
+  RegionBackendService: !ruby/object:Overrides::Terraform::ResourceOverride
```
This is duplicated from the above `BackendService` as they share a nearly identical structure

rileykarson replied:
It's my experience that this is often untrue. While load balancing Regional / Global resources will share a common body, each will use only around 60-70% of the fields.

slevenick commented on third_party/terraform/tests/resource_compute_region_backend_service_test.go.erb:
```diff
@@ -210,24 +210,6 @@ func TestAccComputeRegionBackendService_withSessionAffinity(t *testing.T) {
 	}
 }
 
-func testAccCheckComputeRegionBackendServiceDestroy(s *terraform.State) error {
```
Uses the generated destroy function now

slevenick commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,266 @@ objects:
+                instance group must be within the same region as the
+                BackendService.
+      - !ruby/object:Api::Type::Enum
+        name: 'loadBalancingScheme'
```
This needs to be included in the API call, but must always be `INTERNAL`. Is there a better way to handle this?

rileykarson replied:
Omitting it entirely or using an encoder. This is fine though.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, c0bfb03f93d7b0ad3c7e841e08cd3f3e3cfbee9a. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 9ca6c0b4d002ab762c13b133f49ee97abcea126c. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 579370af48fe5e4ad7f9710f5bfc246996602c8e. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,154 @@ objects:
+          - :CLIENT_IP_PROTO
+          - :CLIENT_IP_PORT_PROTO
+      - !ruby/object:Api::Type::ResourceRef
+        name: 'region'
```
This needs to have an override for `default_from_api` so it's `Computed` + `Optional`

slevenick replied:
Why would region be computed and optional? It's required for a regional backend service

rileykarson replied:
We can pull the region from a few different places, so it ends up being optional: https://github.com/terraform-providers/terraform-provider-google/blob/5ce5686d4997d4b422e63f5e295bc9763f2a997d/google/utils.go#L39-L46

slevenick replied:
Like this?
```
      region: !ruby/object:Overrides::Terraform::PropertyOverride
        default_from_api: true
        required: false
```

rileykarson replied:
Yep! There are some other extra overrides too that make it work a little better as well, see https://github.com/GoogleCloudPlatform/magic-modules/blob/16e6fe42ebdb772e323e3fab7dad503e1de8ea8d/products/compute/terraform.yaml#L52-L58

rileykarson commented on products/compute/terraform.yaml:
```diff
@@ -233,6 +233,46 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+        vars:
+          region_backend_service_name: "region-backend-service"
+          health_check_name: "health-check"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
```
I think most of this shouldn't be here w/o `iap` and `security_policy` right?

slevenick replied:
Right! Removing

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 4b9eed937e734d9b978895206b4f587057aaaa44. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

slevenick commented:
> I think you need to delete the handwritten docs still. For some reason the duplicate generation checking doesn't pick it up, probably because it's copied and not compiled.
> 
> Also, there's a few tests we could make redundant; if https://github.com/terraform-providers/terraform-provider-google-beta/blob/c5816035a0cd37713301bf0edab249ab2a419bd2/google-beta/resource_compute_region_backend_service_test.go#L123-L148 is the example, we can remove that handwritten case entirely. (Old checks against specific fields were used before import-style tests, so they can be ignored generally)

Removed a couple tests, deleted docs. I'm inclined to leave the rest of the tests as they include updates

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 72eb26ba2ab97e67220031735be460234fc24186. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,154 @@ objects:
           How many seconds to wait for the backend before considering it a
           failed request. Default is 30 seconds. Valid range is [1, 86400].
+  - !ruby/object:Api::Resource
+    name: 'RegionBackendService'
```
Let's add [`references`](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/products/compute/api.yaml#L57-L61) to preserve docs like https://github.com/terraform-providers/terraform-provider-google-beta/pull/645/files#diff-83aa3ab4e2ca346150e6962897976dbaL12 for readers

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,154 @@ objects:
+      items: 'items'
+    has_self_link: true
+    description: |
+      A Region Backend Service defines a regionally-scoped group of virtual
```
It's probably worth noting that RBS is only for ILB, like we did before: https://github.com/terraform-providers/terraform-provider-google-beta/pull/645/files#diff-83aa3ab4e2ca346150e6962897976dbaL15

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 16e6fe42ebdb772e323e3fab7dad503e1de8ea8d. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,162 @@ objects:
+      [`google_compute_backend_service`](compute_backend_service.html) instead. 
+
+
+ ## Example Usage
```
```suggestion
 ## Example Usage
```

This seems unintentional

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,162 @@ objects:
+      A Region Backend Service defines a regionally-scoped group of virtual
+      machines that will serve traffic for load balancing.
+
+      ~> **Note**: Region backend services can only be used when using internal load balancing.
```
```suggestion
      Region backend services can only be used when using internal load balancing.
```

The warning box syntax will appear as symbols in most markdown processors, so let's remove it.

rileykarson commented on products/compute/api.yaml:
```diff
@@ -935,6 +935,159 @@ objects:
+
+      Region backend services can only be used when using internal load balancing.
+      For external load balancing, use
+      [`google_compute_backend_service`](compute_backend_service.html) instead. 
```
```suggestion
      a global backend service instead.
```

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 127597d51a4f98de25bc0e34373c67f156b014b1. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 81eb495250a3e2b4ab3518bb868d0c44fe5f7fa9. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 5dd523d7192f1a249212294ef655f5a6604ade40. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, f72104afcfda0e2b53f4b45ecfa079318493388e. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.
